src/support/rand.c

```diff
@@ -132,15 +150,22 @@ __wt_random(WT_RAND_STATE volatile *rnd_state) WT_GCC_FUNC_ATTRIBUTE((visibility
     z = M_Z(rnd);
 
     /*
-     * Check if the value goes to 0 (from which we won't recover), and reset to the initial state.
+     * Check if either of the two value goes to 0 (from which we won't recover), and reset it to the
+     * default initial state. This will never happen with the default seed, but we need this for the
+     * other case.
+     *
+     * We do this one component at a time, so that if the random number generator was initialized
+     * from an explicitly provided seed, it would not reset the entire state and then effectively
+     * result in random number generators from different seeds converging. They would eventually
+     * converge if both W and Z become 0 at the same time, but this is very unlikely.
+     *
      * This has additional benefits if a caller fails to initialize the state, or initializes with a
      * seed that results in a short period.
      */
-    if (z == 0 || w == 0) {
-        __wt_random_init(&rnd);
-        w = M_W(rnd);
-        z = M_Z(rnd);
-    }
+    if (w == 0)
+        M_W(rnd) = w = DEFAULT_SEED_W;
```

Comment:
The assignment to `M_W` is not needed here, only to `w` (and `M_Z` later).  Sometimes redundancy (or extra safety) is warranted, but here I think it works logically to have a single place that the "final" Z and W are adjusted.